# Project 1: Trading with Momentum

## Investment Thesis
Momentum investing, the strategy of buying assets that have recently performed well and selling those that have underperformed, is underpinned by two primary theories: behavioral biases and risk-based explanations. Behavioral finance suggests that cognitive biases, such as investors' tendencies to underreact to new information or to follow herd behavior, lead to price trends that persist over time. These biases cause investors to either overvalue or undervalue assets, creating opportunities for momentum strategies to capitalize on the continuation of these trends. On the other hand, the risk-based perspective posits that momentum profits serve as compensation for bearing additional risk. This view argues that assets exhibiting strong past performance are inherently riskier, and the higher returns associated with momentum strategies are a premium for assuming this increased risk. Both perspectives offer valuable insights into why momentum investing can be effective in financial markets. The code below will demostrate how to implement a momentum strategy and test its significance level to determine whether its a strong or weak trading rule.

## Packages
[Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/) are required. The other packages that we're importing are `helper`, and `project_helper`. These are custom packages built to implement the workflow.  The `helper` and `project_helper` module contains utility functions and graph functions.

### Install Packages

In [ ]:
import sys
!{sys.executable} -m pip install -r requirements.txt

### Load Packages

In [1]:
import pandas as pd
import numpy as np
import yfinance as yf
import datetime
import helper
import project_helper


/Users/mescobars/dev/projects/GitHub/Momentum/.venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


## Market Data
### Load Data
The data we use for most of the projects is end of day data. This contains data for many stocks, but we'll be looking at stocks in the S&P 500. We also made things a little easier to run by narrowing down our range of time period instead of using all of the data.

In [7]:
# Read and print the stock tickers that make up S&P500
tickers = pd.read_html(
    'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies')[0]
print(tickers.head())

# Get the data for this tickers from yahoo finance
start_date = '2019-12-31'
end_date = '2024-12-31'

data = yf.download(tickers.Symbol.to_list(),start_date,end_date, auto_adjust=True)['Close']
#data.head()

[                       1%                       ]  3 of 503 completed

  Symbol             Security             GICS Sector  \
0    MMM                   3M             Industrials   
1    AOS          A. O. Smith             Industrials   
2    ABT  Abbott Laboratories             Health Care   
3   ABBV               AbbVie             Health Care   
4    ACN            Accenture  Information Technology   

                GICS Sub-Industry    Headquarters Location  Date added  \
0        Industrial Conglomerates    Saint Paul, Minnesota  1957-03-04   
1               Building Products     Milwaukee, Wisconsin  2017-07-26   
2           Health Care Equipment  North Chicago, Illinois  1957-03-04   
3                   Biotechnology  North Chicago, Illinois  2012-12-31   
4  IT Consulting & Other Services          Dublin, Ireland  2011-07-06   

       CIK      Founded  
0    66740         1902  
1    91142         1916  
2     1800         1888  
3  1551152  2013 (1888)  
4  1467373         1989  


[*********************100%***********************]  503 of 503 completed

2 Failed downloads:
['BRK.B']: YFTzMissingError('$%ticker%: possibly delisted; no timezone found')
['BF.B']: YFPricesMissingError('$%ticker%: possibly delisted; no price data found  (1d 2019-12-31 -> 2024-12-31)')


### View Data
Run the cell below to see what the data looks like for `close`.

In [8]:
close = data.copy()

### Stock Example
Let's see what a single stock looks like from the closing prices. For this example and future display examples in this project, we'll use Apple's stock (AAPL). If we tried to graph all the stocks, it would be too much information.

In [9]:
apple_ticker = 'AAPL'
project_helper.plot_stock(close[apple_ticker], '{} Stock'.format(apple_ticker))

## Resample Adjusted Prices

The trading signal you'll develop in this project does not need to be based on daily prices, for instance, you can use month-end prices to perform trading once a month. To do this, you must first resample the daily adjusted closing prices into monthly buckets, and select the last observation of each month.

Implement the `resample_prices` to resample `close_prices` at the sampling frequency of `freq`.

In [5]:
def resample_prices(close_prices, freq='M'):
    """
    Resample close prices for each ticker at specified frequency.
    
    Parameters
    ----------
    close_prices : DataFrame
        Close prices for each ticker and date
    freq : str
        What frequency to sample at
        For valid freq choices, see http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
    
    Returns
    -------
    prices_resampled : DataFrame
        Resampled prices for each ticker and date
    """
    # TODO: Implement Function
    prices_resampled = close_prices.resample(freq).last()
    
    return prices_resampled

### View Data
Let's apply this function to `close` and view the results.

In [11]:
monthly_close = resample_prices(close)
monthly_close

Ticker               A        AAPL        ABBV        ABNB         ABT  \
Date                                                                     
2019-12-31   82.442833   71.172134   70.782341         NaN   79.191910   
2020-01-31   79.785271   75.016136   65.645630         NaN   79.784599   
2020-02-29   74.479782   66.411682   69.445663         NaN   70.527962   
2020-03-31   69.389084   61.778706   61.732182         NaN   72.249283   
2020-04-30   74.272110   71.377487   67.595009         NaN   84.671837   
...                ...         ...         ...         ...         ...   
2024-08-31  142.429581  228.748337  192.935364  117.309998  112.150139   
2024-09-30  147.970490  232.743942  194.085236  126.809998  112.882820   
2024-10-31  130.069580  225.661728  201.965424  134.789993  112.778175   
2024-11-30  137.715454  237.330002  181.221054  136.110001  118.149979   
2024-12-31  134.171997  252.199997  174.553925  131.809998  112.211151   

Ticker            ACGL         ACN        ADBE         ADI        ADM  ...  \
Date                                                                   ...   
2019-12-31   40.784039  195.653839  329.809998  108.461227  40.523506  ...   
2020-01-31   41.991680  191.410721  351.140015  100.165085  39.133381  ...   
2020-02-29   38.444828  168.446213  345.119995  100.066452  33.176113  ...   
2020-03-31   27.062571  152.281586  318.239990   82.264626  30.999620  ...   
2020-04-30   22.850092  173.516342  353.640015  100.571167  32.726715  ...   
...                ...         ...         ...         ...        ...  ...   
2024-08-31  107.537109  340.563782  574.409973  232.933716  60.420216  ...   
2024-09-30  106.386528  352.047028  517.780029  229.199524  59.181892  ...   
2024-10-31   93.720558  344.820007  478.079987  222.169296  54.694210  ...   
2024-11-30  100.720001  362.369995  515.929993  217.130630  54.599998  ...   
2024-12-31   91.889999  352.489990  445.799988  212.559998  50.070000  ...   

Ticker             WTW         WY        WYNN        XEL         XOM  \
Date                                                                   
2019-12-31  188.486694  25.053415  135.158691  54.509563   54.752357   
2020-01-31  197.213776  24.016438  122.788376  59.403309   48.742012   
2020-02-29  176.642090  21.552574  105.976105  53.505047   40.941483   
2020-03-31  159.181870  14.232121   59.072994  52.106075   30.220606   
2020-04-30  167.091782  18.363216   83.942726  54.923096   36.985813   
...                ...        ...         ...        ...         ...   
2024-08-31  290.414551  30.301498   76.673813  60.205997  116.978775   
2024-09-30  293.702576  33.650665   95.622864  64.763428  116.264641   
2024-10-31  301.341034  30.967358   95.762482  66.261017  115.828224   
2024-11-30  321.095398  32.259998   94.379997  71.963768  117.959999   
2024-12-31  312.359985  27.889999   85.830002  67.034599  105.760002   

Ticker             XYL         YUM         ZBH        ZBRA         ZTS  
Date                                                                    
2019-12-31   74.241722   91.778297  139.760834  255.440002  127.100456  
2020-01-31   76.946037   96.370384  138.098831  239.020004  129.072998  
2020-02-29   73.100792   81.682961  127.127472  210.970001  128.130478  
2020-03-31   61.560062   62.719696   94.616066  183.600006  113.185310  
2020-04-30   67.958977   79.102051  112.045303  229.660004  124.561760  
...                ...         ...         ...         ...         ...  
2024-08-31  137.146011  134.269379  114.945229  345.380005  182.504501  
2024-09-30  134.653000  139.036285  107.707573  370.320007  194.330627  
2024-10-31  121.439987  130.527512  106.679886  381.970001  178.241150  
2024-11-30  126.750000  138.270004  111.848251  407.000000  174.721786  
2024-12-31  116.279999  133.520004  105.400002  383.850006  161.751007  

[61 rows x 503 columns]

In [ ]:
project_helper.plot_resampled_prices(
    monthly_close.loc[:, apple_ticker],
    close.loc[:, apple_ticker],
    '{} Stock - Close Vs Monthly Close'.format(apple_ticker))

/var/folders/18/blnm1nm50696kqjfj926l25r0000gn/T/ipykernel_99732/3448844271.py:19: FutureWarning:

'M' is deprecated and will be removed in a future version, please use 'ME' instead.



## Compute Log Returns

Compute log returns ($R_t$) from prices ($P_t$) as your primary momentum indicator:

$$R_t = log_e(P_t) - log_e(P_{t-1})$$

Implement the `compute_log_returns` function below, such that it accepts a dataframe (like one returned by `resample_prices`), and produces a similar dataframe of log returns. Use Numpy's [log function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.log.html) to help you calculate the log returns.

In [12]:
def compute_log_returns(prices):
    """
    Compute log returns for each ticker.
    
    Parameters
    ----------
    prices : DataFrame
        Prices for each ticker and date
    
    Returns
    -------
    log_returns : DataFrame
        Log returns for each ticker and date
    """
    # TODO: Implement Function
    log_returns = np.log(prices / prices.shift(1))
    
    return log_returns

### View Data
Using the same data returned from `resample_prices`, we'll generate the log returns.

In [13]:
monthly_close_returns = compute_log_returns(monthly_close)
project_helper.plot_returns(
    monthly_close_returns.loc[:, apple_ticker],
    'Log Returns of {} Stock (Monthly)'.format(apple_ticker))

## Shift Returns
Implement the `shift_returns` function to shift the log returns to the previous or future returns in the time series. For example, the parameter `shift_n` is 2 and `returns` is the following:

```
                           Returns
               A         B         C         D
2013-07-08     0.015     0.082     0.096     0.020     ...
2013-07-09     0.037     0.095     0.027     0.063     ...
2013-07-10     0.094     0.001     0.093     0.019     ...
2013-07-11     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...
```

the output of the `shift_returns` function would be:
```
                        Shift Returns
               A         B         C         D
2013-07-08     NaN       NaN       NaN       NaN       ...
2013-07-09     NaN       NaN       NaN       NaN       ...
2013-07-10     0.015     0.082     0.096     0.020     ...
2013-07-11     0.037     0.095     0.027     0.063     ...
...            ...       ...       ...       ...
```
Using the same `returns` data as above, the `shift_returns` function should generate the following with `shift_n` as -2:
```
                        Shift Returns
               A         B         C         D
2013-07-08     0.094     0.001     0.093     0.019     ...
2013-07-09     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...       ...
...            ...       ...       ...       ...       ...
...            NaN       NaN       NaN       NaN       ...
...            NaN       NaN       NaN       NaN       ...
```
_Note: The "..." represents data points we're not showing._

In [14]:
def shift_returns(returns, shift_n):
    """
    Generate shifted returns
    
    Parameters
    ----------
    returns : DataFrame
        Returns for each ticker and date
    shift_n : int
        Number of periods to move, can be positive or negative
    
    Returns
    -------
    shifted_returns : DataFrame
        Shifted returns for each ticker and date
    """
    # TODO: Implement Function
    shifted_returns = returns.shift(shift_n)
    
    return shifted_returns

### View Data
Let's get the previous month's and next month's returns.

In [15]:
prev_returns = shift_returns(monthly_close_returns, 1)
lookahead_returns = shift_returns(monthly_close_returns, -1)

project_helper.plot_shifted_returns(
    prev_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Previous Returns of {} Stock'.format(apple_ticker))
project_helper.plot_shifted_returns(
    lookahead_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Lookahead Returns of {} Stock'.format(apple_ticker))

## Generate Trading Signal

A trading signal is a sequence of trading actions, or results that can be used to take trading actions. A common form is to produce a "long" and "short" portfolio of stocks on each date (e.g. end of each month, or whatever frequency you desire to trade at). This signal can be interpreted as rebalancing your portfolio on each of those dates, entering long ("buy") and short ("sell") positions as indicated.

Here's a strategy that we will try:
> For each month-end observation period, rank the stocks by _previous_ returns, from the highest to the lowest. Select the top performing stocks for the long portfolio, and the bottom performing stocks for the short portfolio.

Implement the `get_top_n` function to get the top performing stock for each month. Get the top performing stocks from `prev_returns` by assigning them a value of 1. For all other stocks, give them a value of 0. For example, using the following `prev_returns`:

```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0.015     0.082     0.096     0.020     0.075     0.043     0.074
2013-07-09     0.037     0.095     0.027     0.063     0.024     0.086     0.025
...            ...       ...       ...       ...       ...       ...       ...
```

The function `get_top_n` with `top_n` set to 3 should return the following:
```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0         1         1         0         1         0         0
2013-07-09     0         1         0         1         0         1         0
...            ...       ...       ...       ...       ...       ...       ...
```
*Note: You may have to use Panda's [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.iterrows.html) with [`Series.nlargest`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.Series.nlargest.html) in order to implement the function. This is one of those cases where creating a vecorization solution is too difficult.*

In [16]:
def get_top_n(prev_returns, top_n):
    """
    Select the top performing stocks
    
    Parameters
    ----------
    prev_returns : DataFrame
        Previous shifted returns for each ticker and date
    top_n : int
        The number of top performing stocks to get
    
    Returns
    -------
    top_stocks : DataFrame
        Top stocks for each ticker and date marked with a 1
    """
    # TODO: Implement Function
    top_stocks = pd.DataFrame(index=prev_returns.index, columns=prev_returns.columns)
    
    for date, returns in prev_returns.iterrows():
        top_n_returns = returns.nlargest(top_n)
        top_stocks.loc[date, top_n_returns.index] = 1
    
    return top_stocks.fillna(0)

### View Data
We want to get the best performing and worst performing stocks. To get the best performing stocks, we'll use the `get_top_n` function. To get the worst performing stocks, we'll also use the `get_top_n` function. However, we pass in `-1*prev_returns` instead of just `prev_returns`. Multiplying by negative one will flip all the positive returns to negative and negative returns to positive. Thus, it will return the worst performing stocks.

In [17]:
top_bottom_n = 50
df_long = get_top_n(prev_returns, top_bottom_n)
df_short = get_top_n(-1*prev_returns, top_bottom_n)
project_helper.print_top(df_long, 'Longed Stocks')
project_helper.print_top(df_short, 'Shorted Stocks')

10 Most Longed Stocks:
TSLA, ENPH, NVDA, AMD, CRWD, EQT, CZR, MRNA, ALB, TPL
10 Most Shorted Stocks:
CCL, ENPH, WBD, PARA, MRNA, NCLH, HAL, CZR, APA, AMD


/var/folders/18/blnm1nm50696kqjfj926l25r0000gn/T/ipykernel_99732/1914915729.py:24: FutureWarning:

Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`

/var/folders/18/blnm1nm50696kqjfj926l25r0000gn/T/ipykernel_99732/1914915729.py:24: FutureWarning:

Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



## Projected Returns
It's now time to check if your trading signal has the potential to become profitable!

We'll start by computing the net returns this portfolio would return. For simplicity, we'll assume every stock gets an equal dollar amount of investment. This makes it easier to compute a portfolio's returns as the simple arithmetic average of the individual stock returns.

Implement the `portfolio_returns` function to compute the expected portfolio returns. Using `df_long` to indicate which stocks to long and `df_short` to indicate which stocks to short, calculate the returns using `lookahead_returns`. To help with calculation, we've provided you with `n_stocks` as the number of stocks we're investing in a single period.

In [18]:
def portfolio_returns(df_long, df_short, lookahead_returns, n_stocks):
    """
    Compute expected returns for the portfolio, assuming equal investment in each long/short stock.
    
    Parameters
    ----------
    df_long : DataFrame
        Top stocks for each ticker and date marked with a 1
    df_short : DataFrame
        Bottom stocks for each ticker and date marked with a 1
    lookahead_returns : DataFrame
        Lookahead returns for each ticker and date
    n_stocks: int
        The number number of stocks chosen for each month
    
    Returns
    -------
    portfolio_returns : DataFrame
        Expected portfolio returns for each ticker and date
    """
    # TODO: Implement Function
    portfolio_returns = (df_long - df_short) * lookahead_returns / n_stocks
    
    return portfolio_returns

### View Data
Time to see how the portfolio did.

In [19]:
expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)

project_helper.plot_returns(expected_portfolio_returns.T.sum(), 'Portfolio Returns')

## Statistical Tests
### Annualized Rate of Return

In [20]:
expected_portfolio_returns_by_date = expected_portfolio_returns.T.sum().dropna()
portfolio_ret_mean = expected_portfolio_returns_by_date.mean()
portfolio_ret_ste = expected_portfolio_returns_by_date.sem()
portfolio_ret_annual_rate = (np.exp(portfolio_ret_mean * 12) - 1) * 100

print("""
Mean:                       {:.6f}
Standard Error:             {:.6f}
Annualized Rate of Return:  {:.2f}%
""".format(portfolio_ret_mean, portfolio_ret_ste, portfolio_ret_annual_rate))


Mean:                       0.000437
Standard Error:             0.002618
Annualized Rate of Return:  0.53%



The annualized rate of return allows you to compare the rate of return from this strategy to other quoted rates of return, which are usually quoted on an annual basis. 

### T-Test
Our null hypothesis ($H_0$) is that the actual mean return from the signal is zero. We'll perform a one-sample, one-sided t-test on the observed mean return, to see if we can reject $H_0$.

We'll need to first compute the t-statistic, and then find its corresponding p-value. The p-value will indicate the probability of observing a t-statistic equally or more extreme than the one we observed if the null hypothesis were true. A small p-value means that the chance of observing the t-statistic we observed under the null hypothesis is small, and thus casts doubt on the null hypothesis. It's good practice to set a desired level of significance or alpha ($\alpha$) _before_ computing the p-value, and then reject the null hypothesis if $p < \alpha$.

For this project, we'll use $\alpha = 0.05$, since it's a common value to use.

Implement the `analyze_alpha` function to perform a t-test on the sample of portfolio returns. We've imported the `scipy.stats` module for you to perform the t-test.

Note: [`scipy.stats.ttest_1samp`](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.stats.ttest_1samp.html) performs a two-sided test, so divide the p-value by 2 to get 1-sided p-value

In [21]:
from scipy import stats

def analyze_alpha(expected_portfolio_returns_by_date):
    """
    Perform a t-test with the null hypothesis being that the expected mean return is zero.
    
    Parameters
    ----------
    expected_portfolio_returns_by_date : Pandas Series
        Expected portfolio returns for each date
    
    Returns
    -------
    t_value
        T-statistic from t-test
    p_value
        Corresponding p-value
    """
    # TODO: Implement Function
    null_hypothesis = 0.0
    t_statistic, p_value = stats.ttest_1samp(expected_portfolio_returns_by_date, null_hypothesis)
    
    # Divide p-value by 2 for one-tailed test
    p_value /= 2
    
    return t_statistic, p_value

### View Data
Let's see what values we get with our portfolio. After you run this, make sure to answer the question below.

In [22]:
t_value, p_value = analyze_alpha(expected_portfolio_returns_by_date)
print("""
Alpha analysis:
 t-value:        {:.3f}
 p-value:        {:.6f}
""".format(t_value, p_value))


Alpha analysis:
 t-value:        0.167
 p-value:        0.434078



### Question: What p-value did you observe? And what does that indicate about your signal?

*#TODO: Given the observed p-value of 0.43 and a significance level of 0.05, we do not have enough evidence to reject the null hypothesis that the actual mean return from the signal is zero. In other words, the result is not statistically significant at the 5% significance level, so we cannot conclude that the actual mean return from the signal is significantly different from zero. From an investment perspective, the trading signal is not strong enough*